--- a/kpathsea/cnf.c
+++ b/kpathsea/cnf.c
@@ -240,6 +240,22 @@
   string *cnf_files;
   string *cnf;
   const_string cnf_path = kpathsea_init_format (kpse, kpse_cnf_format);
+#define SET_VAR(var, value) \
+  do { \
+    char *tmp1 = xmalloc(sizeof(var)   / sizeof(var[0])   + 1); strcpy(tmp1, var);   \
+    char *tmp2 = xmalloc(sizeof(value) / sizeof(value[0]) + 1); strcpy(tmp2, value); \
+    hash_insert(&(kpse->cnf_hash), tmp1, tmp2); \
+  } while (0)
+
+  SET_VAR("XDG_CONFIG_HOME", "~/.config");
+  SET_VAR("XDG_CACHE_HOME", "~/.cache");
+  SET_VAR("XDG_DATA_HOME", "~/.local/share");
+  SET_VAR("TEXMFVAR", "$XDG_CACHE_HOME/mmtex");
+
+  SET_VAR("TEXMFDOTDIR", ".");
+  SET_VAR("TEXMFUSER", "$XDG_DATA_HOME/{mmtex,texmf}");
+  SET_VAR("TEXMFSYSTEM", "/usr/local/share/{mmtex,texmf}:/usr/share/{mmtex,texmf}");
+  SET_VAR("TEXMF", "{$TEXMFUSER,$TEXMFSYSTEM}");

   cnf_files = kpathsea_all_path_search (kpse, cnf_path, CNF_NAME);
   if (cnf_files && *cnf_files) {
@@ -289,13 +305,6 @@
       free (*cnf);
     }
     free (cnf_files);
-  } else {
-    string warn = getenv ("KPATHSEA_WARNING");
-    if (!(warn && STREQ (warn, "0"))) {
-      WARNING1
-  ("kpathsea: configuration file texmf.cnf not found in these directories: %s",
-        cnf_path);
-    }
   }
 }
 
