--- a/src/lua/lnodelib.c
+++ b/src/lua/lnodelib.c
@@ -1566,7 +1566,7 @@ static int lua_nodelib_direct_setleader(lua_State * L)
 #define get_pdf_literal_direct_value(L,n) do {                  \
     if (pdf_literal_type(n) == lua_refid_literal) {             \
         lua_rawgeti(L, LUA_REGISTRYINDEX, pdf_literal_data(n)); \
-    } else if (pdf_literal_type(n) == lua_refid_literal) {      \
+    } else if (pdf_literal_type(n) == normal) {                 \
         tokenlist_to_luastring(L, pdf_literal_data(n));         \
     }                                                           \
 } while (0)
diff --git a/src/luatex/src/lua/ltexlib.c b/src/luatex/src/lua/ltexlib.c
index 8962a39..eb70c19 100644
--- a/src/lua/ltexlib.c
+++ b/src/lua/ltexlib.c
@@ -1790,7 +1790,7 @@ static int settex(lua_State * L)
                 }
             } else if (is_toks_assign(cur_cmd1)) {
                 if (lua_type(L,i) == LUA_TSTRING) {
-                    j = tokenlist_from_lua(L, -1);  /* uses stack -1 */
+                    j = tokenlist_from_lua(L, i);
                     assign_internal_value((isglobal ? 4 : 0), equiv(cur_cs1), j);
 
                 } else {
